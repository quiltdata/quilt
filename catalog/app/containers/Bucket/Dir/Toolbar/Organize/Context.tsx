import invariant from 'invariant'
import * as React from 'react'

import * as Bookmarks from 'containers/Bookmarks'
import * as Selection from 'containers/Bucket/Selection'
import DeleteDialog, { type DeleteResult } from 'containers/Bucket/Toolbar/DeleteDialog'
import * as Dialogs from 'utils/Dialogs'

interface OrganizeState {
  addToBookmarks: () => void
  removeFromBookmarks: () => void
  toggleBookmarks: () => void
  bookmarkStatus: 'none' | 'partial' | 'all'

  openSelectionPopup: () => void
  clearSelection: () => void

  confirmDeleteSelected: () => void

  selectionCount: number
}

const Context = React.createContext<OrganizeState | null>(null)

function useContext(): OrganizeState {
  const context = React.useContext(Context)
  invariant(context, 'useContext must be used within OrganizeDirProvider')
  return context
}

export const use = useContext

interface OrganizeProviderProps {
  children: React.ReactNode
  onReload: () => void
}

export function OrganizeProvider({ children, onReload }: OrganizeProviderProps) {
  const bookmarks = Bookmarks.use()
  const dialogs = Dialogs.use()
  const slt = Selection.use()

  const handles = React.useMemo(
    () => Selection.toHandlesList(slt.selection),
    [slt.selection],
  )

  const bookmarkStatus = React.useMemo(() => {
    if (!bookmarks || slt.isEmpty) return 'none'
    const bookmarkedCount = handles.filter((handle) =>
      bookmarks.isBookmarked('main', handle),
    ).length
    if (bookmarkedCount === 0) return 'none'
    if (bookmarkedCount === handles.length) return 'all'
    return 'partial'
  }, [bookmarks, slt.isEmpty, handles])

  const addToBookmarks = React.useCallback(() => {
    if (!bookmarks || slt.isEmpty) return
    bookmarks.append('main', handles)
  }, [bookmarks, slt.isEmpty, handles])

  const removeFromBookmarks = React.useCallback(() => {
    if (!bookmarks || slt.isEmpty) return
    bookmarks.remove('main', handles)
  }, [bookmarks, slt.isEmpty, handles])

  const toggleBookmarks = React.useCallback(() => {
    if (bookmarkStatus === 'all') {
      removeFromBookmarks()
    } else {
      addToBookmarks()
    }
  }, [bookmarkStatus, addToBookmarks, removeFromBookmarks])

  const openSelectionPopup = React.useCallback(() => {
    dialogs.open(({ close }) => <Selection.Popup close={close} />)
  }, [dialogs])

  const clearSelection = React.useCallback(() => {
    slt.clear()
  }, [slt])

  const confirmDeleteSelected = React.useCallback(async () => {
    if (slt.isEmpty) return

    const result = await dialogs.open<DeleteResult>(({ close }) => (
      <DeleteDialog close={close} handles={handles} />
    ))

    if (result.deleted) {
      onReload()
      slt.clear()
    }
  }, [dialogs, onReload, slt, handles])

  const selectionCount = slt.totalCount

  const actions = React.useMemo(
    (): OrganizeState => ({
      addToBookmarks,
      removeFromBookmarks,
      toggleBookmarks,
      bookmarkStatus,
      openSelectionPopup,
      clearSelection,
      confirmDeleteSelected,
      selectionCount,
    }),
    [
      addToBookmarks,
      removeFromBookmarks,
      toggleBookmarks,
      bookmarkStatus,
      openSelectionPopup,
      clearSelection,
      confirmDeleteSelected,
      selectionCount,
    ],
  )

  return (
    <Context.Provider value={actions}>
      {children}

      {dialogs.render({ fullWidth: true, maxWidth: 'sm' })}
    </Context.Provider>
  )
}

export { OrganizeProvider as Provider }
