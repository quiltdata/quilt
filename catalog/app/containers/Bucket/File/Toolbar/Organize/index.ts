export * as Context from './Context'
export { default as Options } from './Options'
