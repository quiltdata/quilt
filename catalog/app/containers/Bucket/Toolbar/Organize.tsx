import * as React from 'react'
import * as M from '@material-ui/core'

import { PlaylistAddCheckOutlined as IconPlaylistAddCheckOutlined } from '@material-ui/icons'

import * as Buttons from 'components/Buttons'
import { useSelection } from 'containers/Bucket/Selection/Provider'

const useBadgeClasses = M.makeStyles({
  badge: {
    right: '4px',
  },
})

interface ButtonProps {
  children: NonNullable<React.ReactNode>
  className?: string
  onReload: () => void
}

export default function Button({ onReload, ...props }: ButtonProps) {
  const slt = useSelection()
  const classes = useBadgeClasses()
  return (
    <M.Badge badgeContent={slt.totalCount} classes={classes} color="primary" max={999}>
      <Buttons.WithPopover
        icon={IconPlaylistAddCheckOutlined}
        label="Organize"
        disabled={slt.inited && !slt.totalCount}
        {...props}
      />
    </M.Badge>
  )
}
