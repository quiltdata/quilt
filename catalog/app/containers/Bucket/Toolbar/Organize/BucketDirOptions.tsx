import * as React from 'react'
import * as M from '@material-ui/core'
import {
  TurnedInNotOutlined as IconTurnedInNotOutlined,
  DeleteOutlined as IconDeleteOutlined,
  EditOutlined as IconEditOutlined,
  ClearOutlined as IconClearOutlined,
} from '@material-ui/icons'

import * as Format from 'utils/format'

import * as Context from './ContextDir'

const LIST_ITEM_TYPOGRAPHY_PROPS = { noWrap: true }

interface MenuItemProps {
  className?: string
  icon: React.ReactElement
  primary: string
  onClick: () => void
}

function MenuItem({ className, icon, primary, onClick }: MenuItemProps) {
  return (
    <M.ListItem button onClick={onClick} className={className}>
      <M.ListItemIcon>{icon}</M.ListItemIcon>
      <M.ListItemText
        primary={primary}
        primaryTypographyProps={LIST_ITEM_TYPOGRAPHY_PROPS}
      />
    </M.ListItem>
  )
}

const useStyles = M.makeStyles((t) => ({
  error: {
    color: t.palette.error.main,
  },
}))

export default function BucketDirOptions() {
  const classes = useStyles()
  const {
    addSelectedToBookmarks,
    openSelectionPopup,
    clearSelection,
    confirmDeleteSelected,
    selectionCount,
  } = Context.use()

  return (
    <>
      <M.ListSubheader inset component="div" disableSticky>
        <Format.Plural
          value={selectionCount}
          one="One selected item"
          other={(n) => `${n} selected items`}
        />
      </M.ListSubheader>

      <M.Divider />

      <M.List dense>
        <MenuItem
          icon={<IconTurnedInNotOutlined />}
          onClick={addSelectedToBookmarks}
          primary="Add to bookmarks"
        />
      </M.List>

      <M.Divider />

      <M.List dense>
        <MenuItem
          icon={<IconEditOutlined />}
          onClick={openSelectionPopup}
          primary="Manage selection"
        />
        <MenuItem
          icon={<IconClearOutlined />}
          onClick={clearSelection}
          primary="Clear selection"
        />
      </M.List>

      <M.Divider />

      <M.List dense>
        <MenuItem
          className={classes.error}
          icon={<IconDeleteOutlined color="error" />}
          onClick={confirmDeleteSelected}
          primary="Delete selected items"
        />
      </M.List>
    </>
  )
}
