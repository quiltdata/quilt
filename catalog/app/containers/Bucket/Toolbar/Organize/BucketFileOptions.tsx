import * as React from 'react'
import * as RRDom from 'react-router-dom'
import * as M from '@material-ui/core'
import {
  TurnedInNotOutlined as IconTurnedInNotOutlined,
  TurnedInOutlined as IconTurnedInOutlined,
  DeleteOutlined as IconDeleteOutlined,
  EditOutlined as IconEditOutlined,
  CheckOutlined as IconCheckOutlined,
} from '@material-ui/icons'

import { useConfirm } from 'components/Dialog'
import * as Bookmarks from 'containers/Bookmarks'
import * as FileEditor from 'components/FileEditor'
import * as NamedRoutes from 'utils/NamedRoutes'

import type { FileHandle } from '../types'
import { viewModeToSelectOption } from '../../viewModes'
import type { ViewModes, FileType } from '../../viewModes'

const LIST_ITEM_TYPOGRAPHY_PROPS = { noWrap: true } as const

// NOTE: 'div' is a workaround for hardcoded MUI types
interface MenuItemProps extends Omit<M.ListItemProps<'div'>, 'button'> {
  icon?: React.ReactElement
  children: React.ReactNode
}

function MenuItem({ icon, children, ...props }: MenuItemProps) {
  return (
    <M.ListItem {...props} button>
      {icon && <M.ListItemIcon>{icon}</M.ListItemIcon>}
      <M.ListItemText
        inset={!icon}
        primary={children}
        primaryTypographyProps={LIST_ITEM_TYPOGRAPHY_PROPS}
      />
    </M.ListItem>
  )
}

interface LinkItemProps extends Omit<M.ListItemProps<typeof RRDom.Link>, 'button'> {
  icon?: React.ReactElement
  children: React.ReactNode
}

function LinkItem({ icon, children, ...props }: LinkItemProps) {
  return (
    <M.ListItem {...props} button component={RRDom.Link}>
      <M.ListItemIcon>{icon}</M.ListItemIcon>
      <M.ListItemText
        primary={children}
        primaryTypographyProps={LIST_ITEM_TYPOGRAPHY_PROPS}
      />
    </M.ListItem>
  )
}

interface BookmarksItemProps {
  bookmarks: NonNullable<ReturnType<typeof Bookmarks.use>>
  handle: FileHandle
}

function BookmarksItem({ bookmarks, handle }: BookmarksItemProps) {
  const toggle = React.useCallback(
    () => bookmarks.toggle('main', handle),
    [bookmarks, handle],
  )
  return bookmarks.isBookmarked('main', handle) ? (
    <MenuItem icon={<IconTurnedInOutlined />} onClick={toggle}>
      Remove from bookmarks
    </MenuItem>
  ) : (
    <MenuItem icon={<IconTurnedInNotOutlined />} onClick={toggle}>
      Add to bookmarks
    </MenuItem>
  )
}

interface EditorItemProps {
  editorState: FileEditor.EditorState
}

function EditorItem({ editorState }: EditorItemProps) {
  const handleEditFile = React.useCallback(() => {
    editorState.onEdit(editorState.types[0])
  }, [editorState])
  return (
    <MenuItem icon={<IconEditOutlined />} onClick={handleEditFile}>
      Edit text content
    </MenuItem>
  )
}

interface ViewItemProps {
  handle: FileHandle
  mode: FileType
  selected: boolean
}

function ViewItem({ handle, mode, selected }: ViewItemProps) {
  const { urls } = NamedRoutes.use()
  const { bucket, key, version } = handle
  const label = viewModeToSelectOption(mode).toString()
  return selected ? (
    <MenuItem icon={<IconCheckOutlined />} disabled>
      {label}
    </MenuItem>
  ) : (
    <LinkItem to={urls.bucketFile(bucket, key, { version, mode })}>{label}</LinkItem>
  )
}

const useDeleteItemStyles = M.makeStyles((t) => ({
  root: {
    color: t.palette.error.main,
  },
}))

interface DeleteItemProps {
  handle: FileHandle
}

function DeleteItem({ handle }: DeleteItemProps) {
  const classes = useDeleteItemStyles()
  const confirm = useConfirm({
    title: `Delete "${handle.key}"`,
    submitTitle: 'Delete',
    onSubmit: () => {
      // FIXME: Implement delete logic
      // console.log('Delete file:', handle)
    },
  })
  const handleClick = React.useCallback(
    (event) => {
      event.stopPropagation()
      confirm.open()
    },
    [confirm],
  )
  return (
    <>
      {confirm.render(<></>)}
      <MenuItem
        className={classes.root}
        icon={<IconDeleteOutlined color="error" />}
        onClick={handleClick}
      >
        Delete
      </MenuItem>
    </>
  )
}

const useStyles = M.makeStyles((t) => ({
  error: {
    color: t.palette.error.main,
  },
  subList: {
    '& + &': {
      borderTop: `1px solid ${t.palette.divider}`,
    },
  },
}))

interface BucketFileOptionsProps {
  editorState: FileEditor.EditorState
  handle: FileHandle
  viewModes?: ViewModes
}

export default function BucketFileOptions({
  editorState,
  handle,
  viewModes,
}: BucketFileOptionsProps) {
  const classes = useStyles()
  const bookmarks = Bookmarks.use()

  return (
    <>
      {bookmarks && (
        <M.List dense className={classes.subList}>
          <BookmarksItem bookmarks={bookmarks} handle={handle} />
        </M.List>
      )}

      {FileEditor.isSupportedFileType(handle.key) && (
        <M.List dense className={classes.subList}>
          <EditorItem editorState={editorState} />
        </M.List>
      )}

      {viewModes && viewModes.modes.length > 0 && (
        <M.List
          dense
          subheader={<M.ListSubheader inset>View as</M.ListSubheader>}
          className={classes.subList}
        >
          {viewModes.modes.map((mode) => (
            <ViewItem
              key={mode}
              handle={handle}
              mode={mode}
              selected={mode === viewModes.mode}
            />
          ))}
        </M.List>
      )}

      <M.List dense className={classes.subList}>
        <DeleteItem handle={handle} />
      </M.List>
    </>
  )
}
