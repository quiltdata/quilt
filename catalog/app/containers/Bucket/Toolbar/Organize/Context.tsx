import invariant from 'invariant'
import * as React from 'react'
import * as M from '@material-ui/core'

import Code from 'components/Code'
import * as Bookmarks from 'containers/Bookmarks'
import * as Notifications from 'containers/Notifications'
import * as Model from 'model'
import * as AWS from 'utils/AWS'
import * as Dialogs from 'utils/Dialogs'
import Log from 'utils/Logging'
import * as s3paths from 'utils/s3paths'

import * as Selection from '../../Selection'
import { deleteObject, useFilesListing } from '../../requests'

interface OrganizeDirActions {
  addSelectedToBookmarks: () => void

  openSelectionPopup: () => void
  clearSelection: () => void

  confirmDeleteSelected: () => void

  selectionCount: number
}

const Context = React.createContext<OrganizeDirActions | null>(null)

function useContext(): OrganizeDirActions {
  const context = React.useContext(Context)
  invariant(context, 'useOrganizeDirActions must be used within OrganizeDirProvider')
  return context
}

export const use = useContext

interface DeleteDialogProps {
  close: () => void
  onReload: () => void
}

function DeleteDialog({ close, onReload }: DeleteDialogProps) {
  const slt = Selection.use()
  const [submitting, setSubmitting] = React.useState(false)

  const s3 = AWS.S3.use()
  const { push } = Notifications.use()

  const getFiles = useFilesListing()
  const [selectionHandles, setSelectionHandles] = React.useState<
    Model.S3.S3ObjectLocation[] | null
  >(null)

  React.useEffect(() => {
    async function resolveHandles() {
      const handles = Selection.toHandlesList(slt.selection)
      const filesMap = await getFiles(handles)
      setSelectionHandles(Object.values(filesMap))
    }
    resolveHandles()
  }, [getFiles, slt.selection])

  const onSubmit = React.useCallback(async () => {
    setSubmitting(true)
    const errors: Array<{ handle: Model.S3.S3ObjectLocation; error: any }> = []
    let successCount = 0

    if (selectionHandles) {
      await Promise.all(
        selectionHandles.map(async (handle) => {
          try {
            await deleteObject({ s3, handle })
            successCount++
          } catch (error) {
            Log.error('Failed to delete object:', error)
            errors.push({ handle, error })
          }
        }),
      )
    }

    if (successCount > 0) {
      push(`Successfully deleted ${successCount} file${successCount !== 1 ? 's' : ''}`)
    }

    if (errors.length > 0) {
      const errorDetails = errors
        .slice(0, 3)
        .map(({ handle }) => s3paths.handleToS3Url(handle))
        .join(', ')
      const moreText = errors.length > 3 ? ` and ${errors.length - 3} more` : ''

      push(
        `Failed to delete ${errors.length} file${errors.length !== 1 ? 's' : ''}: ${errorDetails}${moreText}`,
      )
    }

    setSubmitting(false)
    slt.clear()
    close()
    onReload()
  }, [close, s3, slt, push, onReload, selectionHandles])

  return (
    <>
      <M.DialogTitle>Delete selected items</M.DialogTitle>
      <M.DialogContent>
        {selectionHandles ? (
          <M.List dense disablePadding>
            {selectionHandles.map(({ bucket, key, version }) => (
              <M.ListItem key={`${bucket}${key}${version}`} disableGutters>
                <M.ListItemText
                  primary={
                    <Code>
                      s3://{bucket}/{key}
                    </Code>
                  }
                  secondary={version}
                />
              </M.ListItem>
            ))}
          </M.List>
        ) : (
          <M.CircularProgress size={64} />
        )}
      </M.DialogContent>
      <M.DialogActions>
        <M.Button onClick={close} color="primary" variant="outlined">
          Cancel
        </M.Button>
        <M.Button
          color="primary"
          disabled={submitting}
          variant="contained"
          onClick={onSubmit}
        >
          Submit
        </M.Button>
      </M.DialogActions>
    </>
  )
}

interface OrganizeDirProviderProps {
  children: React.ReactNode
  onReload: () => void
}

export function OrganizeDirProvider({ children, onReload }: OrganizeDirProviderProps) {
  const bookmarks = Bookmarks.use()
  const dialogs = Dialogs.use()
  const slt = Selection.use()

  const addSelectedToBookmarks = React.useCallback(() => {
    if (!bookmarks || slt.isEmpty) return
    bookmarks.append('main', Selection.toHandlesList(slt.selection))
  }, [bookmarks, slt.isEmpty, slt.selection])

  const openSelectionPopup = React.useCallback(() => {
    dialogs.open(({ close }) => <Selection.Popup close={close} />)
  }, [dialogs])

  const clearSelection = React.useCallback(() => {
    slt.clear()
  }, [slt])

  const confirmDeleteSelected = React.useCallback(async () => {
    if (slt.isEmpty) return
    dialogs.open(({ close }) => <DeleteDialog onReload={onReload} close={close} />)
  }, [dialogs, onReload, slt])

  const selectionCount = slt.totalCount

  const actions = React.useMemo(
    (): OrganizeDirActions => ({
      addSelectedToBookmarks,
      openSelectionPopup,
      clearSelection,
      confirmDeleteSelected,
      selectionCount,
    }),
    [
      addSelectedToBookmarks,
      openSelectionPopup,
      clearSelection,
      confirmDeleteSelected,
      selectionCount,
    ],
  )

  return (
    <Context.Provider value={actions}>
      {children}

      {dialogs.render({ fullWidth: true, maxWidth: 'sm' })}
    </Context.Provider>
  )
}

export { OrganizeDirProvider as Provider }
