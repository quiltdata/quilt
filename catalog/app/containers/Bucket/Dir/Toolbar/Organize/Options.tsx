import * as React from 'react'
import * as M from '@material-ui/core'
import * as Icons from '@material-ui/icons'

import * as Format from 'utils/format'
import assertNever from 'utils/assertNever'

import * as Context from './Context'

const LIST_ITEM_TYPOGRAPHY_PROPS = { noWrap: true }

interface MenuItemProps {
  className?: string
  icon: React.ReactElement
  primary: string
  onClick: () => void
}

function MenuItem({ className, icon, primary, onClick }: MenuItemProps) {
  return (
    <M.ListItem button onClick={onClick} className={className}>
      <M.ListItemIcon>{icon}</M.ListItemIcon>
      <M.ListItemText
        primary={primary}
        primaryTypographyProps={LIST_ITEM_TYPOGRAPHY_PROPS}
      />
    </M.ListItem>
  )
}

const useStyles = M.makeStyles((t) => ({
  error: {
    color: t.palette.error.main,
  },
}))

export default function OrganizeOptions() {
  const classes = useStyles()
  const {
    toggleBookmarks,
    bookmarkStatus,
    openSelectionPopup,
    clearSelection,
    confirmDeleteSelected,
    selectionCount,
  } = Context.use()

  const bookmarkIcon = React.useMemo(() => {
    switch (bookmarkStatus) {
      case 'all':
        return <Icons.TurnedInOutlined />
      case 'partial':
        return <Icons.BookmarksOutlined />
      case 'none':
        return <Icons.TurnedInNotOutlined />
      default:
        return assertNever(bookmarkStatus)
    }
  }, [bookmarkStatus])

  const bookmarkText = React.useMemo(() => {
    switch (bookmarkStatus) {
      case 'all':
        return 'Remove from bookmarks'
      case 'partial':
        return 'Add all to bookmarks'
      case 'none':
        return 'Add to bookmarks'
      default:
        return assertNever(bookmarkStatus)
    }
  }, [bookmarkStatus])

  return (
    <>
      <M.ListSubheader inset component="div" disableSticky>
        <Format.Plural
          value={selectionCount}
          one="One selected item"
          other={(n) => `${n} selected items`}
        />
      </M.ListSubheader>

      <M.Divider />

      <M.List dense>
        <MenuItem icon={bookmarkIcon} onClick={toggleBookmarks} primary={bookmarkText} />
      </M.List>

      <M.Divider />

      <M.List dense>
        <MenuItem
          icon={<Icons.EditOutlined />}
          onClick={openSelectionPopup}
          primary="Manage selection"
        />
        <MenuItem
          icon={<Icons.ClearOutlined />}
          onClick={clearSelection}
          primary="Clear selection"
        />
      </M.List>

      <M.Divider />

      <M.List dense>
        <MenuItem
          className={classes.error}
          icon={<Icons.DeleteOutlined color="error" />}
          onClick={confirmDeleteSelected}
          primary="Delete selected items"
        />
      </M.List>
    </>
  )
}
