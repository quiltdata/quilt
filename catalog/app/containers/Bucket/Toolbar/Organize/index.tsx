import * as React from 'react'
import * as M from '@material-ui/core'

import { PlaylistAddCheckOutlined as IconPlaylistAddCheckOutlined } from '@material-ui/icons'

import * as Buttons from 'components/Buttons'

import { useSelection } from '../../Selection/Provider'

import * as Context from './Context'

export { default as BucketDirOptions } from './BucketDirOptions'
export { default as BucketFileOptions } from './BucketFileOptions'

const useBadgeClasses = M.makeStyles({
  badge: {
    right: '4px',
  },
})

interface ButtonProps {
  children: NonNullable<React.ReactNode>
  className?: string
  label?: string
}

export function Button(props: ButtonProps) {
  const slt = useSelection()
  const classes = useBadgeClasses()
  return (
    <Context.Provider>
      <M.Badge badgeContent={slt.totalCount} classes={classes} color="primary" max={999}>
        <Buttons.WithPopover
          icon={IconPlaylistAddCheckOutlined}
          label="Organize"
          disabled={slt.inited && !slt.totalCount}
          {...props}
        />
      </M.Badge>
    </Context.Provider>
  )
}
