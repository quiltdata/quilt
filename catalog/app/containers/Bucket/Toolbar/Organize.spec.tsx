import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Organize from './Organize'

jest.mock('@material-ui/core', () => ({
  ...jest.requireActual('@material-ui/core'),
  IconButton: ({ className, children, onClick }: any) => (
    <button role="button" {...{ className, children, onClick }} />
  ),
  Button: ({ className, children, onClick }: any) => (
    <button role="button" {...{ className, children, onClick }} />
  ),
}))

jest.mock('containers/Bucket/Selection/Provider', () => ({
  useSelection: jest.fn(() => ({
    isEmpty: false,
    totalCount: 5,
    inited: true,
  })),
}))

describe('containers/Bucket/Toolbar/Organize', () => {
  const mockOnReload = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should render with selection items', () => {
    const { container } = render(
      <Organize onReload={mockOnReload} className="custom-class">
        Hello, Popover!
      </Organize>,
    )
    expect(container).toMatchSnapshot()
  })

  it('should render children when popover is opened', () => {
    const { container, getByRole } = render(
      <Organize onReload={mockOnReload}>Hello, Popover!</Organize>,
    )
    const button = getByRole('button')
    act(() => {
      fireEvent.click(button)
    })
    expect(container).toMatchSnapshot()
  })
})
