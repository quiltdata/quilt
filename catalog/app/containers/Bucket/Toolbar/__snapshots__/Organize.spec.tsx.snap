// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Bucket/Toolbar/Organize should render children when popover is opened 1`] = `
<div>
  <span
    class="MuiBadge-root"
  >
    <div
      class="makeStyles-root-7"
    >
      <button
        role="button"
      >
        Organize
      </button>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root makeStyles-backdrop-9"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        class="MuiPaper-root makeStyles-popup-8 MuiPaper-elevation4 MuiPaper-rounded"
      >
        Hello, Popover!
      </div>
    </div>
    <span
      class="MuiBadge-badge makeStyles-badge-6 MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
    >
      5
    </span>
  </span>
</div>
`;

exports[`containers/Bucket/Toolbar/Organize should render with selection items 1`] = `
<div>
  <span
    class="MuiBadge-root"
  >
    <div
      class="makeStyles-root-2"
    >
      <button
        class="custom-class"
        role="button"
      >
        Organize
      </button>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root makeStyles-backdrop-4"
        style="opacity: 0; visibility: hidden;"
      />
    </div>
    <span
      class="MuiBadge-badge makeStyles-badge-1 MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
    >
      5
    </span>
  </span>
</div>
`;
